### PR TITLE
Interactive markers

### PR DESCRIPTION
**Pull request by sofroniewn**
# What does this PR do? Why are doing this? References?
This PR builds on functionality initially developed in #73 to support the creation and deletion of markers. In order to make use of the markers interactively a new boolean property has been added to the viewer - `viewer.interactive`. If this flag is set to true. And a marker layer is visible, selected, and above the highest image layer, then new markers will be created by clicking on the image where there are currently no markers, and an existing marker will be deleted by clicking on it.

There a couple things to note at this point. First do we like having an interactive flag on the viewer? I felt like if clicking on the image when a marker layer was selected was sufficient to create a new marker then we should at least have a little barrier to starting it. If we like this idea there should probably both be a keyboard short cut and a toggle button on the gui to enable editing. Alternatively we could use a modifier key like shift to "enter interactive mode" and enable editing.

Right now I don't support dragging of markers, this should be easy to add, and I'll probably do it pretty soon, but given the ease of creating / deleting markers right now didn't seem as important.

Should the total number of markers be shown and updated in the markers widget properties window? Probably yes, I just havn't added it yet.

A major outstanding question is should there be a way to select markers - (i.e. click selects, then shift /cntrl click selects more or drag a selection around selects even more). Could these then be recolored / resized or moved together as a group independently of the other markers in the layer? In this case something like double click would be needed for delete. Open to suggestions.

image

## Type of change
[Please delete options that are not relevant.]
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
[Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.]
- [x] TestCase Run examples.layers.ipynb and interact with gui

## Final Checklist:
- [x] My PR is the possible minimum work for the desired functionality

### CODE REVIEW

**Comment by sofroniewn** [comment]
another example

image

**Comment by freeman-lab** [comment]
This is awesome! Really showing the promise of napari-gui, and already a better UX than a lot of similar tools! I pulled the PR and played around for awhile, and chatted offline with Nick a bit, wanted to summarize the key takeaways here.

The main UX challenge in the current version is that there's no visual feedback about what actions I can take, and similar click behaviors do interactive and non-interactive things. Specifically, I often found myself getting stuck in the following ways:
- clicking and not adding points and forgetting why (didn't set interactive to true? in the wrong layer?)
- clicking to pan and worrying I would accidentally add a point
- clicking to add or delete point but accidentally dragging and panning

This all suggests one easy solution. To make annotation mode more explicit, we can have a hot key that enables / disables it, and while enabled, the **cursor changes** to indicate what's possible.

Specifically, after you press "A", if you are:
- hovering over empty space 
   -> cursor turns into a `+`, click creates a new mark
- hovering over an existing mark 
   -> cursor turns into an `x`, click deletes the mark
- hovering over an existing mark while also holding shift
   -> cursor turns into the move sign, click and drag moves the mark

If you press "A" again: 
- hovering anywhere
   -> cursor is the default pointer, you cannot interact with or modify marks

In other words, pressing "A" toggles "annotation" mode on or off, and the cursor changes accordingly. Nick found this [link](http://doc.qt.io/qt-5/qcursor.html) re: modifying the cursor in qt, seems relatively straightforward to use the existing options or roll our own. And here's a nice [D3 example](https://bl.ocks.org/mbostock/3750558) of cursor changes working practice. 

Note: This is more or less copying the concept from Adobe of having different tools activated by hotkeys, but trying to keep it as simple and streamlined as possible.

There are a few details to work out in this proposal, but what do others think of the basic idea?

**Comment by jni** [comment]
Great stuff @sofroniewn and great suggestions @freeman-lab! (And absolutely love the example screenshot! :joy:) The only thing I would change is making the default move, and shift-click being delete.

**Comment by jni** [comment]
I would also argue that we are now getting into something where a plugin *would* be the right thing. For example, as a long term goal, it would be great for "A" mode to generate code to add the clicked points to a Python list.

Not sure we need to stop you in your tracks for this particular vision, just wanted to flag it since we are starting to have this discussion over at https://github.com/Napari/napari-core/issues/26

**Comment by sofroniewn** [comment]
I have switched the flag from `viewer.interactive` to `viewer.annotation` and added a button to the UI that could can click to enter annotation mode. We can also add a hotkey `a` when we add hotkeys, but as we don't have any in place right now I didn't want to start in this PR. We can also move the button to somewhere else in the future or change it's styling, but for now I think it's ok.

I really like the idea from @freeman-lab to change the cursor when we're in annotation mode to be more helpful. I will make a + when nothing is selected and over empty image for adding a marker, a hand/move symbol when nothing is selected and over a marker for dragging, and a delete symbol when shift is down and over a marker.

One thing that is still unclear to me is when in annotation mode if you attempt to drag when not over a marker do I just register a mouse press or mouse release and create a new marker at either of those (if so which one) or do I perform a pan action as is done right now. I feel like I still like having panning enabled when in annotation mode otherwise you have to keep switching in and out as you want to move around the image.

@jni right now the markers that you create do get added to a list in the `.data` (and `.coords`) property of the layer so they are instantly available to you in the notebook. I'm not quite sure what you mean by

> it would be great for "A" mode to generate code to add the clicked points to a Python list

I think basic interactivity / annotation like this is something that should be built into the viewer and supported for markers, lines, polygons etc., maybe even a paintbrush like tool to do pixel level annotations. I think things that then use those annotations (like to do foreground / background segmentation as is done in tools like ilastik) should be custom plugins.

**Comment by sofroniewn** [comment]
So I'm playing with the changing cursor now and it make apparent one behavior that is very strange - if a multiple (say two) layers are selected then clicking empty space will create markers in both right now. Clicking on a spot that has markers in both will delete markers in both. Clicking in a spot that has a marker in one but not the other will delete the existing marker in the layer that had one and then create a new marker in the layer that didn't. In this case what should the cursor look like when it hovers over this point? Should it be a plus or a delete? Or is the whole idea of being able to edit multiple marker layers simultaneously confusing and should it be prohibited. Curious what everyone thinks - I know @freeman-lab ran into this a little.

Also if you are in annotation mode but have no markers layers selected then should the cursor just look normal, as in this case clicking won't create a new marker. I think this behavior makes the more sense than annotation mode somehow enforcing that there is a marker layer selected.

**Comment by sofroniewn** [comment]
Have added a preliminary version of the cursor change - I used default cursors from [here](http://doc.qt.io/qt-5/qcursor.html#a-note-for-x11-users) but I don't like them very much. Will probably add my own soon. You can see the weird behavior that I mentioned in the above comment. Not sure what to do about it.

I also still havn't implemented the dragging of markers nor made deletion with a shift. Thought I'd just give this a try first.

**Comment by jni** [comment]
They're so retro! Back in the Windows 95 days I installed a Super Mario World-themed cursor set. (The hourglass was Mario's walking animation.) It was *great*. I propose we do the same here. Then you can have Wario as the cursor when something crazy is about to happen. =P

More seriously, I suggest that we never edit more than one layer at once, at least initially. It just sounds terrible to me even *before* you consider the opposite-toggle dilemma. Editing a single layer should also make the code a lot simpler.

Regarding panning, 100% I think it should remain on when clicking and dragging. That should make it really nice UX to annotate lots of points in big images.

Finally, about my cryptic code generation comment: one of my long term goals for Napari is to blur the line between interactive work and Python code. Specifically, *anything other than visualisation* should be recorded as a Python script that you can run later to get the exact same result. In terms of interacting with the notebook etc, I would aim for each state-modifying action to be run in the same IPython kernel as Napari, so that your IPython history log can be re-run later, minus the Napari GUI calls, and you get the same annotations back.

Whether we implement this interactivity as a plugin or as a built-in, we should at least start to think about recording stuff now that we are touching data and not just visualisation.

Does that clarify things a little?

**Comment by freeman-lab** [comment]
Hehe I think I had the same cursor set @jni :)

I'm definitely on board with limiting editing to one layer at once. In general I'm all for guardrails that make things easy to understand for the majority of users, and only limit rarer power use cases, which we can study further as they arise and potentially develop custom solutions for.

On the panning topic...

> I feel like I still like having panning enabled when in annotation mode otherwise you have to keep switching in and out as you want to move around the image.

> Regarding panning, 100% I think it should remain on when clicking and dragging. That should make it really nice UX to annotate lots of points in big images.

I agree, in theory :) My concern is the minor hiccup where you go to click and it doesn't register an add marker because you accidentally dragged a tiny bit so it instead registers as a pan. To me, if we have a hot key, switching in and out of annotation mode is still quite fluid. In Illustrator, for example, you have to hold spacebar to pan, but it becomes quite natural to do that alongside drawing, and prevents any potential ambiguity in interpreting the click. 

In our case, switching in and out of annotation mode is currently "stateful", so it's not quite as streamlined, but there's a minor variant where 
- holding "A" puts you in annotation mode temporarily
- tapping "A" twice switches to it 

(think shift vs caps-lock). With that, you could fluidly swap between marking and panning by holding and releasing "A" (if you are annotating lots of points across a big image), or hit "A" twice (if you are doing a bunch of marking in one image region). Of course, the "A" button in the UI only gives you the on/off style because you can't click in two places at once.

Is anyone else experiencing the kind of glitch I'm describing? Happy to sit down together and show you @sofroniewn , maybe I'm imagining it :)

**Comment by sofroniewn** [comment]
Hmm on  limiting editing to one layer at a time it's not 100% clear to me how we want to do this in the UI, I guess just making marker edits apply ONLY to the top selected & visible marker layer could work, so if multiple marker layers are selected only the top one is getting changed and the cursor changes are only reflecting what is or is not present in the top selected one.

Not sure I like the double clicking A hotkey idea - seems to confusing for me. Having panning off while in annotation mode and then holding spacebar to enable panning (and change the cursor temporarily) could work - it is pretty natural and fast.

**Comment by sofroniewn** [comment]
One more reason to support the pan with spacebar in a fixed annotation mode is to support another "paintbrush" style annotation I have planned. For this annotation layer clicking or dragging the cursor will paint on a new layer ontop of the image (essentially making a binary mask for pixelwise accurate segmentations). In this case dragging is required for continuous painting and so cannot be used for panning, but you will still want to easily pan and so I see holding the spacebar as the only reasonable compromise.

**Comment by freeman-lab** [comment]
@sofroniewn awesome idea, spacebar for pan while in annotation mode seems like a great solution (and consistent with other tools, e.g. Adobe). I'd feel free to leave it out of this PR, as we may want to tackle support for key mapping in an entire separate PR. So for now pan will just be disabled while annotating, but we'll get it back soon.

**Comment by sofroniewn** [comment]
I have updated the PR to support drag and drop of the markers when in annotation mode. When in annotation mode new markers are added by clicking (multiple markers can now be added to the same spot). Markers can be dragged by holding down shift and dragging. Markers can be deleted by control clicking. Both pan and zoom are disabled.

I have not implemented the spacebar to turn them back on yet, and think that hotkeys can come in a future PR. We'll have to decide whether to use vispy keyboard events or the native qt events. The same sort of holds true for mouse events. Now I'm using a mix of both, which simplifies some things, but is more clunky in other ways.

Now that I have shift click for dragging, we could just leave pan and zoom enabled for normal interactivity, but there can be this thing that @freeman-lab noted where you click to create a new marker but maybe drag a little or something. This could actually be dealt with too. But the main reason I want to leave it is the planned annotation layer I mentioned before that will be like a paint-brush where this won't be an option. Happy still to reconsider too.

I still havn't implemented the cursor changes, will do them next.

There's still no concept of multiple markers being selected and their properties being editable as a group independent of the others. I think I'd still like to see this done, but it can wait for a future PR too.

**Comment by freeman-lab** [comment]
nice! tried it out, the use of keys is feeling good, only weirdness: shift clicking in an empty space (not on an existing mark) creates a new one, that seems like not the correct behavior

**Comment by sofroniewn** [comment]
I fixed this shift click problem - that was a bug.

I have also added a hotkey `a` that toggles entering / exiting annotation mode. And pressing and holding space bar when in annotation mode to reenable zoom/pan and turn off annotation. How I have implemented the hotkeys is pretty crude and we might want to pull this functionality out to a separate place where it is all contained, but that can come in a future PR.

Adding / moving / deleting markers now applies to only the top selected layer.

I have tried to implement changing the cursor using pyqt's default cursors but the results are buggy at best. I am pretty confident I have the logic correct, but most of the time the cursor does unpredictable things and can undergo state changes simply but swapping in and out of the application. I am pretty unsatisfied with the results right now, but am curious if people have similar problems on their computers. Searching for help online so far has not yielded much. If the cursor problem was solved I'd say the PR was ready to merge aside from any comments on implementation details.

File: gui/elements/_viewer.py [hunk 1]
```diff
@@ -306,5 +345,99 @@ def update_statusBar(self):
...
+            elif event.key == 'Shift':
+                if self.annotation and self._active_markers:
+                    self._qt.canvas.native.setCursor(Qt.PointingHandCursor)
+            elif event.key == 'Meta':
+                if self.annotation and self._active_markers:
+                    self._qt.canvas.native.setCursor(Qt.ForbiddenCursor)
+            elif event.key == 'a':
+                cb = self.layers._qt.layersControls.annotationCheckBox
+                cb.setChecked(not cb.isChecked())
+
```
Comment (AhmetCanSolak) [inline gui/elements/_viewer.py:425]: This is such a hacky way to implement the desired functionality. What about switching these if elseif structures to QActions and setting shortcuts to these actions?

Comment (sofroniewn) [reply gui/elements/_viewer.py:425]: sure, happy to change. Right now I'm using vispy to detect the keypress events and there's actually a bug where if focus if shifted to another widget the keypress events won't be detected. clicking back in the viewer restores focus to vispy and they work again. i think @jni noticed this behavior when he was playing around.

@AhmetCanSolak how do you feel about the need to fix this now vs after we merge making a separate PR that cleans up all keyboard shortcuts and adds in some new ones like ctrl+w to close? I definitely don't see what I have here as a long term solution.

Comment (AhmetCanSolak) [reply gui/elements/_viewer.py:425]: it is okay to not solve right away IMO, we can let this PR hang here and can merge later after fixing. Or you can comment out this part only and later on these lines will let us to remember the intention and we can merge the PR and have a follow-up PR.

Comment (kne42) [reply gui/elements/_viewer.py:425]: @AhmetCanSolak what about this implementation makes you think that it’s hacky?

**Comment by sofroniewn** [comment]
I have update the PR based on some feedback at the last thursday meeting:

I have made the layer name now appear in the bottom left during mouse over events instead of the layer index. Currently layer names need not be unique, but that could be enforced in a later PR.

I've added some helpful text about enabling pan/zoom when in annotation mode to the bottom right of the status bar. Right now that's being done with one widget and a show / hide. If we think we want to have different text there in different contexts then we might want to do this in a more elegant way, but I think that can come in a future PR when needed.

I've tried to implement a custom cursor image when no marker layers are selected in annotation mode so it is not confusing when no markers are created after a click. It required some pretty horrible linking across aspects of the gui, but I think it's important functionality. During our refactor I think we'll want to think more about how different parts of the gui get linked and communicate. Maybe making better use of events. Also right now sometimes the cursor does not have the right image and sometime it changes unexpectedly after I leave and return to the window. I think these are PyQt problems that I havn't figured out. If anyone else has any experience with this would love to hear.

People requested at the meeting that only Pan and not Zoom gets disabled when entering annotation mode. I don't know how to do this in vispy, and a quick look at the docs didn't indicate an easy option. Does anyone here know how to do it? Could ask the vispy team as well, but I'd probably say we can press on without it and then fix later too.

Finally, I reproduced a bug that @jni had found where the hot-key to toggle annotation was stops working if focus shifts out of the viewer to another part of the gui. This is probably because I'm using vispy for the keypress events. If instead we attached the keypress event to the whole window and used Qt we wouldn't have this problem. I think this makes sense in the long run, but I think that sort of refactoring can come in a followup PR.

In summary I think the PR is ready to be merged, despite the caveats, and any pending addressing any additional comments from the team.

**Review by royerloic — APPROVED** [review]

**Review by AhmetCanSolak — APPROVED** [review]

**Review by kne42 — APPROVED** [review]
